bevy_lint/tests/ui.rs

```diff
@@ -55,3 +71,55 @@ fn config() -> color_eyre::Result<Config> {
 
     Ok(config)
 }
+
+#[derive(Deserialize, Debug)]
+#[serde(rename = "compiler-artifact", tag = "reason")]
+struct ArtifactMessage<'a> {
+    package_id: &'a str,
+    target: ArtifactTarget<'a>,
+    filenames: Vec<&'a Path>,
+}
+
+#[derive(Deserialize, Debug)]
+struct ArtifactTarget<'a> {
+    name: &'a str,
+    kind: Vec<&'a str>,
+}
+
+fn find_bevy_rlib() -> color_eyre::Result<PathBuf> {
+    let output = Command::new("cargo")
+        .arg("build")
+        .arg("--test=ui")
+        .arg("--message-format=json")
+        .stderr(Stdio::inherit())
+        .output()?;
+
+    ensure!(output.status.success());
+
+    const NEWLINE: u8 = '\n' as u8;
+    const BEVY_PACKAGE_ID_PREFIX: &str =
+        "registry+https://github.com/rust-lang/crates.io-index#bevy@";
```

Comment:
Not really knowing about how the package IDs work, it looks like this uses the Github registry, is this compatible with the new sparse registry? Or is it completely independent of that?

Reply:
That's a good point! I searched through the documentation and found [the specification for package IDs](https://doc.rust-lang.org/cargo/reference/pkgid-spec.html).

> The following are references to the `regex` package on `crates.io`:
>
> | Spec                                                              | Name    | Version |
> |:------------------------------------------------------------------|:-------:|:-------:|
> | `regex`                                                           | `regex` | `*`     |
> | `regex@1.4`                                                       | `regex` | `1.4.*` |
> | `regex@1.4.3`                                                     | `regex` | `1.4.3` |
> | `https://github.com/rust-lang/crates.io-index#regex`              | `regex` | `*`     |
> | `https://github.com/rust-lang/crates.io-index#regex@1.4.3`        | `regex` | `1.4.3` |
> | `registry+https://github.com/rust-lang/crates.io-index#regex@1.4.3` | `regex` | `1.4.3` |

It does not look like there's a difference between the Git-based and sparse registry types, they all use the same format.

While reading the spec, I did see this, though:

> The specification may be fully qualified, such as `https://github.com/rust-lang/crates.io-index#regex@1.4.3` or it may be abbreviated, such as `regex`. The abbreviated form may be used as long as it uniquely identifies a single package in the dependency graph. If there is ambiguity, additional qualifiers can be added to make it unique. For example, if there are two versions of the regex package in the graph, then it can be qualified with a version to make it unique, such as `regex@1.4.3`.

In summary, there can be multiple forms of package IDs. Because Cargo does not guarantee that it will start with `registry+https://github.com/rust-lang/crates.io-index#bevy@`, we probably shouldn't be depending on that. (It also prevents us from changing the dependency to be a file path or a Git repository, which is another reason to remove it.)